### PR TITLE
use DEFERRABLE constraints to clean the data after tests

### PR DESCRIPTION
**Pull request by xcoulon**
Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: gormsupport/cleaner/db_clean.go [hunk 1]
```diff
@@ -36,17 +36,14 @@ import (
...
 		keys  map[string]interface{}
 	}
 	var entities []entity
-	hookRegistered := db.Callback().Create().Get(hookName) != nil
-	if hookRegistered {
-		hookName += "-" + uuid.NewV4().String()
-	}
-	db.Callback().Create().After("gorm:create").Register(hookName, func(scope *gorm.Scope) {
+	// hook to monitor all created entities
+	createHookName := newHookname()
```
Comment (alexeykazakov) [inline gormsupport/cleaner/db_clean.go]: createHookName is a bit confusing here. Shouldn't we call it `hookName`?

File: gormsupport/cleaner/db_clean.go [hunk 2]
```diff
@@ -55,25 +52,48 @@ func DeleteCreatedEntities(db *gorm.DB) func() {
...
 		entities = append(entities, entity{table: scope.TableName(), keys: keys})
 	})
+
 	return func() {
-		defer db.Callback().Create().Remove(hookName)
+		// drop all constraints
+
+		defer func() {
+			db.Callback().Create().Remove(createHookName)
+			// re-create all constraints
```
Comment (alexeykazakov) [inline gormsupport/cleaner/db_clean.go]: You are not re-create all constraints here. You are using another defer for that.

File: gormsupport/cleaner/db_clean.go [hunk 3]
```diff
@@ -55,25 +52,48 @@ func DeleteCreatedEntities(db *gorm.DB) func() {
...
 		}
+		// defer all DB constraints that can be deferred (see https://www.postgresql.org/docs/8.2/sql-set-constraints.html)
+		_, err := tx.CommonDB().Exec("SET CONSTRAINTS ALL DEFERRED")
+		if err != nil {
+			log.Error(nil, map[string]interface{}{
+				"error": err,
+			}, "failed to defer all constraints before cleaning the test records in the DB")
+		}
+		defer func() {
+			_, err := tx.CommonDB().Exec("SET CONSTRAINTS ALL DEFERRED")
```
Comment (alexeykazakov) [inline gormsupport/cleaner/db_clean.go]: `SET CONSTRAINTS ALL IMMEDIATE` I guess?

File: gormsupport/cleaner/db_clean.go [hunk 4]
```diff
@@ -55,25 +52,48 @@ func DeleteCreatedEntities(db *gorm.DB) func() {
...
+
 		for i := len(entities) - 1; i >= 0; i-- {
 			entity := entities[i]
 			log.Debug(nil, map[string]interface{}{
-				"table":     entity.table,
-				"keys":      entity.keys,
-				"hook_name": hookName,
+				"table": entity.table,
+				"keys":  entity.keys,
+				// "hook_name": hookName,
```
Comment (alexeykazakov) [inline gormsupport/cleaner/db_clean.go]: Why not logging it?

File: gormsupport/cleaner/db_clean.go [hunk 5]
```diff
@@ -55,25 +52,48 @@ func DeleteCreatedEntities(db *gorm.DB) func() {
...
 			}, "Deleting entities from '%s' table with keys %v", entity.table, entity.keys)
 			if len(entity.keys) == 0 {
 				log.Panic(nil, map[string]interface{}{
-					"table":     entity.table,
-					"hook_name": hookName,
+					"table": entity.table,
+					// "hook_name": hookName,
```
Comment (alexeykazakov) [inline gormsupport/cleaner/db_clean.go]: Why not logging it?

**Comment by xcoulon** [comment]
thanks for the review @alexeykazakov, and let me update the PR, as it contains intermediate changes that were not meant to remain in such a form :/

File: gormsupport/cleaner/db_clean.go (hunk 5 above)
Comment (xcoulon) [reply gormsupport/cleaner/db_clean.go]: I reverted that change

File: gormsupport/cleaner/db_clean.go (hunk 2 above)
Comment (xcoulon) [reply gormsupport/cleaner/db_clean.go]: yes, it was a left over comment. I fixed that

File: gormsupport/cleaner/db_clean.go (hunk 1 above)
Comment (xcoulon) [reply gormsupport/cleaner/db_clean.go]: yes, I reverted that

**Comment by xcoulon** [comment]
@sbryzak I've just updated the PR, it's in a more reviewable state now ;)

File: gormsupport/cleaner/db_clean.go [hunk 6]
```diff
@@ -55,14 +52,36 @@ func DeleteCreatedEntities(db *gorm.DB) func() {
...
+			}
+		}()
+		// defer all DB constraints that can be deferred (see https://www.postgresql.org/docs/8.2/sql-set-constraints.html)
+		_, err := tx.CommonDB().Exec("SET CONSTRAINTS ALL DEFERRED")
+		if err != nil {
+			log.Error(nil, map[string]interface{}{
+				"error": err,
+			}, "failed to defer all constraints before cleaning the test records in the DB")
+		}
+		defer func() {
```
Comment (alexeykazakov) [inline gormsupport/cleaner/db_clean.go]: Another leftover :)

Comment (xcoulon) [reply gormsupport/cleaner/db_clean.go]: doh!

**Pull request merged**